[css-anchor-position-1] Use Style::ScopedName to represent anchor names

tuankiet65 opened the pull request:
#### 696b0afcef81c484ec7c878812f211a52fb3d191
<pre>
[css-anchor-position-1] Use Style::ScopedName to represent anchor names
<a href="https://rdar.apple.com/139617595">rdar://139617595</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=282914">https://bugs.webkit.org/show_bug.cgi?id=282914</a>

Reviewed by Antti Koivisto.

Currently, AtomString is used to represent anchor names in the code.
However, the anchor names are supposed to be scoped (i.e: anchor names
in one shadow tree can only be referred to in the same shadow tree).
To prepare for this, this patch mechanically replace all uses of AtomString
to Style::ScopedName. Note that this patch does not fix logic to consider
scopes when resolving anchor names, that&apos;ll come in a separate patch.

* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::valueForAnchorName):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/css/calc/CSSCalcTree+Evaluation.cpp:
(WebCore::CSSCalc::evaluate):
(WebCore::CSSCalc::evaluateWithoutFallback):
* Source/WebCore/css/calc/CSSCalcTree+Parser.cpp:
(WebCore::CSSCalc::consumeAnchor):
* Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
(WebCore::CSSCalc::simplify):
* Source/WebCore/css/calc/CSSCalcTree.h:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::anchorNames const):
(WebCore::RenderStyle::initialAnchorNames):
(WebCore::RenderStyle::initialPositionAnchor):
(WebCore::RenderStyle::positionAnchor const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setAnchorNames):
(WebCore::RenderStyle::setPositionAnchor):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::AnchorPositionEvaluator::findAnchorAndAttemptResolution):
(WebCore::Style::AnchorPositionEvaluator::evaluate):
(WebCore::Style::AnchorPositionEvaluator::evaluateSize):
(WebCore::Style::collectAnchorsForAnchorName):
(WebCore::Style::AnchorPositionEvaluator::updateSnapshottedScrollOffsets):
* Source/WebCore/style/AnchorPositionEvaluator.h:
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertAnchorName):
(WebCore::Style::BuilderConverter::convertPositionAnchor):

Canonical link: <a href="https://commits.webkit.org/286841@main">https://commits.webkit.org/286841@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/388bfb4ded3772b447aae904a0e4b445dcb0cff3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77205 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56240 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/30120 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81768 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28487 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79322 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65388 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4536 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60492 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18539 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80272 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50454 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66269 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40792 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47857 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23767 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26810 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68969 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24102 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83192 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4585 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3093 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68764 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/content-visibility-auto-shared-element.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4741 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66242 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68021 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16994 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12010 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10104 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4532 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/7347 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4551 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7986 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6310 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/style/AnchorPositionEvaluator.h:
```diff
@@ -50,17 +50,20 @@ enum class AnchorPositionResolutionStage : uint8_t {
 
-using AnchorElements = HashMap<AtomString, WeakRef<Element, WeakPtrImplWithEventTargetData>>;
+using AnchorName = ScopedName;
+using AnchorNameOrAuto = std::optional<ScopedName>;
```
I find these aliases to add more confusion/inconsistency than clarification personally. I'd rather use ScopedName directly here.

Not to mention the extra unnecessary includes.

tuankiet65 replied:
I got rid of the aliases and just use `ScopedName` or `std::optional<ScopedName>` as appropriate.

nt1m commented on Source/WebCore/rendering/style/RenderStyle.h:
```diff
@@ -2239,13 +2240,13 @@ class RenderStyle final : public CanMakeCheckedPtr<RenderStyle> {
-    static Vector<AtomString> initialAnchorNames();
-    inline const Vector<AtomString>& anchorNames() const;
-    inline void setAnchorNames(const Vector<AtomString>&);
+    static Vector<Style::AnchorName> initialAnchorNames();
```
Nit: `Vector<Style::ScopedName>` isn't the most effective structure, since the scopeOrdinal that's being provided for each item is the same in the whole vector. View Transition classes has the same issue, so we should probably fix it there too.

Maybe we should have: 
```
struct Style::ScopedNameVector {
    Vector<AtomString> names;
    ScopeOrdinal scopeOrdinal;
}
```

anttijk commented on Source/WebCore/style/ScopedName.h:
```diff
@@ -47,3 +48,30 @@ inline WTF::TextStream& operator<<(WTF::TextStream& ts, ScopedName scopedName)
+    static unsigned hash(const ScopedName& key) { return computeHash(key.name); }
+
+    // FIXME: compare the scope ordinal too.
+    static bool equal(const ScopedName& a, const ScopedName& b) { return a.name == b.name; }
```
why not do these things already in this patch?

tuankiet65 replied:
We had a discussion about mechanically replacing things with `ScopedName` and you mentioned using the name hash as the hash, but yeah we could just hash the whole thing.

anttijk replied:
It is probably nicer to either not have traits at all (and just keep hashing on the AtomString manually) or have proper traits.

As mentioned, I'm not sure how useful making these hashable is going to be in long run. The ordinals need to be resolved and replaced with scope roots before they can actually be compared for equality/looked up from a global hash.

tuankiet65 replied:
Is there a reason why we don't put the scope root pointer into `ScopedName`? Instead of having to resolve to the scope root when it's needed?

tuankiet65 committed:
[css-anchor-position-1] Use Style::ScopedName to represent anchor names
rdar://139617595
https://bugs.webkit.org/show_bug.cgi?id=282914

Reviewed by Antti Koivisto.

Currently, AtomString is used to represent anchor names in the code.
However, the anchor names are supposed to be scoped (i.e: anchor names
in one shadow tree can only be referred to in the same shadow tree).
To prepare for this, this patch mechanically replace all uses of AtomString
to Style::ScopedName. Note that this patch does not fix logic to consider
scopes when resolving anchor names, that'll come in a separate patch.

* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::valueForAnchorName):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/css/calc/CSSCalcTree+Evaluation.cpp:
(WebCore::CSSCalc::evaluate):
(WebCore::CSSCalc::evaluateWithoutFallback):
* Source/WebCore/css/calc/CSSCalcTree+Parser.cpp:
(WebCore::CSSCalc::consumeAnchor):
* Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
(WebCore::CSSCalc::simplify):
* Source/WebCore/css/calc/CSSCalcTree.h:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::anchorNames const):
(WebCore::RenderStyle::initialAnchorNames):
(WebCore::RenderStyle::initialPositionAnchor):
(WebCore::RenderStyle::positionAnchor const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setAnchorNames):
(WebCore::RenderStyle::setPositionAnchor):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::AnchorPositionEvaluator::findAnchorAndAttemptResolution):
(WebCore::Style::AnchorPositionEvaluator::evaluate):
(WebCore::Style::AnchorPositionEvaluator::evaluateSize):
(WebCore::Style::collectAnchorsForAnchorName):
(WebCore::Style::AnchorPositionEvaluator::updateSnapshottedScrollOffsets):
* Source/WebCore/style/AnchorPositionEvaluator.h:
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertAnchorName):
(WebCore::Style::BuilderConverter::convertPositionAnchor):

Canonical link: https://commits.webkit.org/286841@main
